---
title: "Untitled"
author: "Lu Zimeng"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(Rsubread)
  library(rtracklayer)
  library(Biostrings)
  library(Rfastp)
  library(epiwraps)
})
ah <- AnnotationHub()
```
```{r, eval=FALSE}
options(timeout=600) 
dir.create("raw")
download.file("https://www.encodeproject.org/files/ENCFF127RRR/@@download/ENCFF127RRR.fastq.gz", dest="raw/Myc.fastq.gz", mode = "wb")
```
# Alignment
## Using Rsubread
### Building a genome index for mapping
```{r, eval=FALSE}
genome <- ah[["AH49674"]]
dir.create("BDGP6_genome")
export(import.2bit(genome), "BDGP6_genome/genome.fasta.gz", compress=TRUE)
Rsubread::buildindex("BDGP6_genome/rsubread", reference="BDGP6_genome/genome.fasta.gz")
```

### Alignment
```{r}
dir.create("aligned")
align.stats <- Rsubread::align(index="BDGP6_genome/rsubread", type="dna",
                                readfile1="rfastp.trimmed/Myc_R1.fastq.gz",
                               output_file="aligned/Myc.bam",
                               nthreads=6, sortReadsByCoordinates=TRUE)
align.stats
```
#### a) how many reads (and what percentage) was mapped: 7446729 (46.2%)

# Peak calling
## Using R
```{r}
peaks <- callPeaks("aligned/Myc.bam", fragLength=50L)
```
#### b) how many peaks were found: 4222
```{r}
length(peaks)
```
#### c) Plot the signal around one of the peaks
```{r}
BiocManager::install("rtracklayer")
```

```{r}
head(peaks)
region <- as.character(granges(peaks[1]))
plotSignalTracks(c(Myc="aligned/Myc.bam"), region=region)
plotSignalTracks(list(signal="aligned/Myc.bam"), region=region)
```




Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
