---
title: "assignment"
author: "Lu Zimeng"
date: "`r Sys.Date()`"
output: html_document
---
```{r}
suppressPackageStartupMessages({
  library(epiwraps)
  library(AnnotationHub)
  library(MotifDb)
  library(universalmotif)
  library(ensembldb)
  library(ggplot2)
})
ah <- AnnotationHub(localHub=TRUE)
ensdb <- ah[["AH89211"]] # mouse ensembldb object
```
## Download the data

```{r, eval=FALSE}
options(timeout=1800)
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam", "atac.chr19.bam", mode="wb")
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam.bai", "atac.chr19.bam.bai", mode="wb")
```
## Preparing tracks

```{r}
bam = "atac.chr19.bam"
# create a track using only nucleosome-free fragments, the number of cuts/insertion sites at each position
bam2bw(bam, output_bw = "NF_cuts.bw", paired=TRUE, binWidth=1L, type="ends", minFragLength=30, 
       maxFragLength=120, shift=c(4L,-5L), forceSeqlevelsStyle = "Ensembl")
# create a track using nucleosome-containing fragments, the number of cuts/insertion sites at each position
bam2bw(bam, output_bw = "N_cuts.bw", paired=TRUE, binWidth=1L, type="ends", minFragLength=140, 
       shift=c(4L,-5L), forceSeqlevelsStyle = "Ensembl")
```
## Obtaining the sites with a CTCF motif

```{r}
# get CTCF motif
motif <- query(MotifDb, c("CTCF","Mus"))[[1]]
motif2 <- convert_motifs(motif, class="TFBSTools-PFMatrix")
genome <- ah[["AH68356"]]
# get the sequence for chr19:
chr19 <- import(genome)["19"]
# find motif occurences
moi <- motifmatchr::matchMotifs(motif2, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges (when using motifmatchr)
moi <- as(setNames(moi,names(chr19)), "GRanges")
```
## Plotting signal around the motif occurences

```{r}
# we prepare the list of tracks
tracks <- c("N cuts"="N_cuts.bw", "NF cuts"="NF_cuts.bw")
sm <- signal2Matrix(tracks, moi, w=5, extend=500)
# plot the signals:
plotEnrichedHeatmaps(sm, trim=0.95)
# we renormalize using border backgrounds
nf <- bwNormFactors(tracks, useSeqLevels="19", nwind=5000L)
smb <- renormalizeBorders(sm)
plotEnrichedHeatmaps(smb, trim=0.95, colors = c("white","darkred"), minRowVal = 20)
```
## Signal around MAZ
```{r}
# fetch the protein-coding MAZ from chromosome 19
download.file("https://www.encodeproject.org/files/ENCFF058WWE/@@download/ENCFF058WWE.bed.gz", "mESC_MAZ_ENCFF058WWE.bed.gz")
maz <- rtracklayer::import("mESC_MAZ_ENCFF058WWE.bed.gz", format="NarrowPeak")
seqlevelsStyle(maz) <- "Ensembl"
maz.sm <- signal2Matrix(tracks, granges(maz), w=5, extend=3000)
maz.smb <- rescaleSignalMatrices(maz.sm, nf)
plotEnrichedHeatmaps(maz.smb, trim=0.95, colors=c("white","darkred"))
```
## Signal around FOSL1
```{r}
# fetch the protein-coding FOSL1 from chromosome 19
download.file("https://www.encodeproject.org/files/ENCFF536WWR/@@download/ENCFF536WWR.bed.gz", "mESC_KLF4_ENCFF536WWR.bed.gz")
klf4 <- rtracklayer::import("mESC_KLF4_ENCFF536WWR.bed.gz", format="NarrowPeak")
seqlevelsStyle(klf4) <- "Ensembl"
klf4.sm <- signal2Matrix(tracks, granges(klf4), w=5, extend=3000)
klf4.smb <- rescaleSignalMatrices(klf4.sm, nf)
plotEnrichedHeatmaps(klf4.smb, trim=0.95, colors=c("white","darkred"))
```



---